### DISCUSSION TITLE
Get root entity from Select in SQLAlchemy 2.0

### DISCUSSION THREAD

**Discussion by michaelbukachi** [sqlalchemy/sqlalchemy#7872] [Usage Questions]
Is it possible to get the root object of a query using SQLAlchemy 2.0? It used to be possible to do so 1.3. Here's a snippet I used to use to get the root entity from a query
```python
def _get_root_cls(query: Query) -> "Type[Base]":
    if query._from_obj:
        return get_mapper(query._from_obj[0]).class_ 

    if hasattr(query, "_entity_from_pre_ent_zero"):
        return query._entity_from_pre_ent_zero().class_

    raise ValueError("Cannot get a root class from`{}`".format(query))
```
With SQAlchemy we have the `Select` object, however, I can't figure out how to achieve the same thing.

**Comment by zzzeek** [comment]
hi -

the use is the same with legacy Query as well as Select() which is to use [column_descriptions](https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=column_descriptions#sqlalchemy.orm.Query.column_descriptions):

* 1.x: https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=column_descriptions#sqlalchemy.orm.Query.column_descriptions
* 2.0: https://docs.sqlalchemy.org/en/14/orm/queryguide.html#queryguide-inspection

**Reply by michaelbukachi** [reply]
Thanks for the links. Can't believe I've never seen those links. It doesn't seem to work though. Take a look at the following snippet:
```
query = select(func.count(), func.count(Car.id))\
        .select_from(Owner)\
        .join(Owner.cars)

print(query)
```
gives
```
[{'name': 'count', 'type': Integer(), 'aliased': False, 'expr': <sqlalchemy.sql.functions.count at 0x7ff46b118f10; count>, 'entity': None}, {'name': 'count_1', 'type': Integer(), 'aliased': False, 'expr': <sqlalchemy.sql.functions.count at 0x7ff46930c7f0; count>, 'entity': <class 'app.db.tests.test_utils.Car'>}]

```
I was expecting `Owner` to be in the list based on `select_from`. However, this is not the case.

**Reply by zzzeek** [reply]
oh, we dont have public API for that right now.   I just added such a method to DML for Insert/Update/Delete so perhaps we should define some accessor called "entity_descriptions" for this.

take a look at 9a8d95ce05a88c1efb02b52a88ff35 where I've just added some new accessors.    One that derives from `_from_obj` can be added also.

**Reply by michaelbukachi** [reply]
Nice. The addition would be great. Right now the workaround I'm using is passing around the entity objects together with the query whenever I need to extract some info.

**Reply by zzzeek** [reply]
what do you do when the query is against a join ?

**Reply by michaelbukachi** [reply]
It works as long as the query is selecting a whole model or the first column being selected is from the right entity it works. Btw here's an update to the `_get_root_class`:
```python
def _get_root_cls(query: Select) -> "Type[Base]":
    if query.columns_clause_froms:
        if isinstance(query.columns_clause_froms[0], Table):
            return get_mapper(query.columns_clause_froms[0]).class_

    if hasattr(query, "_raw_columns"):
        return get_mapper(getattr(query, "_raw_columns")[0]).class_

    if hasattr(query, "_from_obj"):
        return get_mapper(getattr(query, "_from_obj")[0]).class_

    if hasattr(query, "_entity_from_pre_ent_zero"):
        return getattr(query, "_entity_from_pre_ent_zero")().class_

    raise ValueError("Cannot get a root class from`{}`".format(query))

```
However, it doesn't fully work.
```python
# models
class Model:
    other = relationship('Other')

class Other:
    id = Column()

# This works
select(Model)
# And this works
select(Model.id)

# However this does not
# _get_root_cls returns Other instead of Model
select(func.count(), func.count(Other.id)).select_from(Model),join(Model.other)

```

**Reply by zzzeek** [reply]
im looking for what would you expect the "root entity" is for this query:

     stmt = select(func.count('*')).select_from(join(join(A, B), C))

or even what is the list of entities.  they are nested in a join structure.   this is why features like this are not easy to add.

**Reply by michaelbukachi** [reply]
Hmm. I see what you mean. Since `select_from` has a wide variety of allowed parameters it makes it difficult to derive the "root entity"

**Reply by zzzeek** [reply]
i would likely add a function people can use to just get the entitty from the table.

it's not changing anytime soon so a simplistic version of that would be

```py
def entity_from_table(table):
    return table._annotations.get("parententity", None)
```

**Reply by michaelbukachi** [reply]
What about a function to determine all tables involved in a query? :thinking:

**Reply by zzzeek** [reply]
you can use [visitors.traverse()](https://docs.sqlalchemy.org/en/14/core/visitors.html?highlight=visitors%20traverse#sqlalchemy.sql.visitors.traverse) for that.   this is not often API people on the outside use but it should be pretty solid.

**Reply by michaelbukachi** [reply]
Interesting. I've tried:
```python
query = select(func.count(), func.count(Car.id)).select_from(Owner).join(Owner.cars)
def visit_table(table):
    print(table)
visitors.traverse(query, {}, {'table': visit_table})
```
It seems the last entry is `owner`. So the entity in `from` section of a query always appears last?

**Reply by zzzeek** [reply]
i dont think visitors.traverse can give any guarantees about ordering.   for an ordered traverse you'd need to do that manually, like select._from_obj and all that as you are doing.   visitors.traverse can still be used for sub-elements of that construct.

**Reply by michaelbukachi** [reply]
I'll have to play around with different queries in our codebase to see if orderings are different. We do have a standard way of creating queries though so hopefully the results are consistent.
